plugins/utils/yanky.nix

```diff
@@ -6,17 +6,132 @@
   ...
 }:
 with lib;
-let
-  cfg = config.plugins.yanky;
-in
-{
-  options.plugins.yanky = helpers.neovim-plugin.extraOptionsOptions // {
-    enable = mkEnableOption "yanky.nvim";
+helpers.neovim-plugin.mkNeovimPlugin config {
+  name = "yanky";
+  originalName = "yanky.nvim";
+  defaultPackage = pkgs.vimPlugins.yanky-nvim;
+
+  maintainers = [ maintainers.GaetanLepage ];
+
+  # TODO: introduced 2024-06-28. Remove after 24.11 release.
+  deprecateExtraOptions = true;
+  optionsRenamedToSettings = [
+    [
+      "ring"
+      "historyLength"
+    ]
+    [
+      "ring"
+      "storage"
+    ]
+    [
+      "ring"
+      "storagePath"
+    ]
+    [
+      "ring"
+      "syncWithNumberedRegisters"
+    ]
+    [
+      "ring"
+      "cancelEvent"
+    ]
+    [
+      "ring"
+      "ignoreRegisters"
+    ]
+    [
+      "ring"
+      "updateRegisterOnCycle"
+    ]
+    [
+      "picker"
+      "telescope"
+      "useDefaultMappings"
+    ]
+    [
+      "systemClipboard"
+      "syncWithRing"
+    ]
+    [
+      "highlight"
+      "onPut"
+    ]
+    [
+      "highlight"
+      "onYank"
+    ]
+    [
+      "highlight"
+      "timer"
+    ]
+    [
+      "preserveCursorPosition"
+      "enabled"
+    ]
+    [
+      "textobj"
+      "enabled"
+    ]
+  ];
+  imports =
+    let
+      basePluginPath = [
+        "plugins"
+        "yanky"
+      ];
+    in
+    [
+      (mkRemovedOptionModule
+        (
+          basePluginPath
+          ++ [
+            "picker"
+            "select"
+            "action"
+          ]
+        )
+        ''
+          Please use `plugins.yanky.settings.picker.select.action` instead.
+          However, be careful as we do not perform any manipulation on the provided string.
+        ''
+      )
+      (mkRenamedOptionModule (
+        basePluginPath
+        ++ [
+          "picker"
+          "telescope"
+          "enable"
+        ]
+      ) (basePluginPath ++ [ "enableTelescope" ]))
+      (mkRemovedOptionModule
+        (
+          basePluginPath
+          ++ [
+            "picker"
+            "telescope"
+            "mappings"
+          ]
+        )
+        ''
+          Please use `plugins.yanky.settings.picker.telescope.mappings` instead.
+          However, be careful as we do not perform any manipulation on the provided strings.
+        ''
+      )
+    ];
 
-    package = helpers.mkPluginPackageOption "yanky.nvim" pkgs.vimPlugins.yanky-nvim;
+  settingsDescription = ''
+    Options provided to the `require('yanky').setup` function.
 
+    NOTE: The following local variables are available in the scope:
+    ```lua
+    local utils = require('yanky.utils')
+    local mapping = require('yanky.telescope.mapping') -- Only if `plugins.yanky.enableTelescope` is `true`
+    ```
```

Comment:
This is fine, but I wonder if splitting it up (something like this) would be better?

```md
NOTE: The variable `utils = require('yanky.utils')` is made available to this scope.\
The variable `mapping = require('yanky.telescope.mapping')` is also made available when `plugins.telescope` is enabled.
```

or

```md
NOTE:
- The variable `utils = require('yanky.utils')` is made available to this scope.
- The variable `mapping = require('yanky.telescope.mapping')` is made available when `plugins.telescope` is enabled.
```

Reply:
Just throwing out ideas, I'm fine with whatever so long as it is readable and accurate